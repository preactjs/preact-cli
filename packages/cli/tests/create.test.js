const { access, readFile } = require('fs').promises;
const { join } = require('path');
const { create } = require('./lib/cli');
const { snapshotDir } = require('./lib/utils');
const shell = require('shelljs');
const dirTree = require('directory-tree');

describe('preact create', () => {
	it('scaffolds the `default` official template', async () => {
		let dir = await create('default');

		const directoryTree = dirTree(dir, {
			exclude: /node_modules|package-lock|yarn.lock/,
		});

		// Creating a stable name, as the test directory is normally a randomized string
		directoryTree.name = 'default-template-project';

		expect(await snapshotDir([directoryTree], false)).toMatchSnapshot();
	});

	it('should use template.html from the github repo', async () => {
		let dir = await create('netlify');
		const template = await readFile(join(dir, 'src/template.html'), 'utf8');
		expect(template.includes('twitter:card')).toEqual(true);
	});

	describe('CLI Options', () => {
		it('--name', async () => {
			let dir = await create('simple', { name: 'renamed' });
			const packageJson = await readFile(join(dir, 'package.json'), 'utf8');

			expect(JSON.parse(packageJson).name).toBe('renamed');

			// @ts-ignore
			const mockExit = jest.spyOn(process, 'exit').mockImplementation(() => {});
			await create('simple', { name: '*()@!#!$-Invalid-Name' });
			expect(mockExit).toHaveBeenCalledWith(1);
			mockExit.mockRestore();
		});

		it('--git', async () => {
			let dir = await create('simple', { git: true });
			expect(await access(join(dir, '.git'))).toBeUndefined();

			dir = await create('simple', { git: false });
			await expect(access(join(dir, '.git'))).rejects.toThrow(
				'no such file or directory'
			);
		});

		it('--invalid-arg', () => {
			const { code, stderr } = shell.exec(
				`node ${join(__dirname, '../src/index.js')} create --invalid-arg`
			);
			expect(stderr).toMatch(
				"Invalid argument '--invalid-arg' passed to create."
			);
			expect(code).toBe(1);
		});
	});
});
