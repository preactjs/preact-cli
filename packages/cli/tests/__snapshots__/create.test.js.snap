// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`preact create scaffolds the \`default\` official template 1`] = `
"default-template-project
  .gitignore
  README.md
  package.json
  src
    assets
      favicon.ico
      icons
        android-chrome-192x192.png
        android-chrome-512x512.png
        apple-touch-icon.png
        favicon-16x16.png
        favicon-32x32.png
        mstile-150x150.png
      preact-logo-inverse.svg
      preact-logo.svg
    components
      app.js
      header
        index.js
        style.css
    index.js
    manifest.json
    routes
      home
        index.js
        style.css
      profile
        index.js
    style
      index.css
    sw.js
    template.html
  tests
    __mocks__
      browserMocks.js
      fileMocks.js
      setupTests.js
    header.test.js
"
`;
